src/zk_proof_systems/ppzksnark/r1cs_ppzksnark/r1cs_ppzksnark.tcc

```diff
@@ -758,5 +758,221 @@ bool r1cs_ppzksnark_affine_verifier_weak_IC(const r1cs_ppzksnark_verification_ke
+    //computing the random coefficients that will be used
+    auto r_1 = Fr<ppT>::random_element();
+    auto r_2 = Fr<ppT>::random_element();
+    auto r_3 = Fr<ppT>::random_element();
+    auto r_4 = Fr<ppT>::random_element();
+    auto r_5 = Fr<ppT>::random_element();
+
+
+    //computing left input for the first ML factor
+    // r1Pi_a and vk_A
+    acc.pair1 = acc.pair1 + r_1*proof.g_A.g;
+    
+    //computing left input for the second ML factor
+    // r1Pi'_a + R2Pi'_B+r3Pi'_C + r5Pi_C and -g2
+    acc.pair2 = acc.pair2 + r_1*proof.g_A.h+r_2*proof.g_B.h + r_3*proof.g_C.h + r_5*proof.g_C.g;
+    //computing left input for the third ML factor
+    // r3Pi_c and vk_C
+    acc.pair3 = acc.pair3 + r_3*proof.g_C.g;
+
+    //computing left input for the fourth ML factor
+    // r4Pi_K and vk_gamma
+    acc.pair4 = acc.pair4 + r_4*proof.g_K;
+        //computing left input for the fifth ML factor
+    //−r 4(vk x + πA + πC) and vk^2_betagamma
+    acc.pair5 = acc.pair5 + -r_4*(accu + proof.g_A.g + proof.g_C.g);
+    
+    //computing left input for the six ML factor
+    //r5Pi_H and -vk_Z
+    acc.pair6 = acc.pair6 + r_5*proof.g_H;
+    
+    //computing left input for the seventh ML factor
+    //r_2 vk_B-r_4 vk^1_{\beta\gamma}+r_5(vk_x + \pi_A) and pi_B
+    auto left_7 = ppT::precompute_G1(r_2*vk.alphaB_g1-r_4*vk.gamma_beta_g1+r_5*(accu+proof.g_A.g));
+    auto right_7 = ppT::precompute_G2(proof.g_B.g);
+    acc.pair7 = acc.pair7*ppT::miller_loop(left_7,right_7);
+
+    leave_block("Call to r1cs_ppzksnark_batcher");
+}
+
+
+
+
+/*A verifier that batches all pairing checks into one using random coefficients
+  requires only one FE (final exponentiation).
+  Also requires only one call to multiple_miller_loop, as only needs to compute one product of 
+  Miller loops (MLs).
+  Also uses rule e(a,c)*e(b,c)=e(a+b,c) to reduce number of factors in the ML product
+  If proof is wrong has at most 1/|F| probability of accepting it
+  Always accepts correct proofs
+*/
+template<typename ppT>
+bool r1cs_ppzksnark_probabilistic_verifier(const r1cs_ppzksnark_verification_key<alt_bn128_pp> &vk,
+                                            const r1cs_ppzksnark_primary_input<alt_bn128_pp> &primary_input,
+                                            const r1cs_ppzksnark_proof<alt_bn128_pp> &proof)
+{
+    enter_block("Call to r1cs_ppzksnark_probabilistic_verifier");
+    const accumulation_vector<G1<ppT> > accumulated_IC = vk.encoded_IC_query.template accumulate_chunk<Fr<ppT> >(primary_input.begin(), primary_input.end(), 0);
+    const G1<ppT> &acc = accumulated_IC.first;
+    //computing the random coefficients that will be used
+    auto r_1 = Fr<ppT>::random_element();
+    auto r_2 = Fr<ppT>::random_element();
+    auto r_3 = Fr<ppT>::random_element();
+    auto r_4 = Fr<ppT>::random_element();
+    auto r_5 = Fr<ppT>::random_element();
+
+    enter_block("Preparing 7 ML factors");
+
+    //computing left input for the first ML factor
```

Comment:
I think it's fine as is, in this context it's clear right is not correct. Also, I'm using first,second,.. for the ML factors